politeiawww/paywall.go

```diff
@@ -1,47 +1,280 @@
 package main
 
 import (
-	www "github.com/decred/politeia/politeiawww/api/v1"
+	"fmt"
+	"sync"
+	"time"
+
+	v1 "github.com/decred/politeia/politeiawww/api/v1"
 	"github.com/decred/politeia/politeiawww/database"
 	"github.com/decred/politeia/util"
 )
 
-// ProcessVerifyUserPaymentTx verifies that the provided transaction
+type paywallInfo struct {
+	address     string
+	amount      uint64
+	txNotBefore int64
+}
+
+const (
+	// paywallExpiry is the amount of time the server will watch a paywall address
+	// for transactions. It gets reset when the user logs in or makes a call to
+	// RouteVerifyUserPayment.
+	paywallExpiry = time.Hour * 24
+
+	// paywallCheckGap is the amount of time the server sleeps after polling for
+	// a paywall address.
+	paywallCheckGap = time.Second * 1
+)
+
+var (
+	paywallUsers map[uint64]paywallInfo
+	mutex        sync.RWMutex
+	once         sync.Once
+)
+
+func paywallHasExpired(txNotBefore int64) bool {
+	expiryTime := time.Unix(txNotBefore, 0).Add(paywallExpiry)
+	return time.Now().After(expiryTime)
+}
+
+func initPaywallUsersPool() {
+	if paywallUsers == nil {
+		paywallUsers = make(map[uint64]paywallInfo)
+	}
+}
+
+func addUser(user *database.User) {
+	initPaywallUsersPool()
+	paywallUsers[user.ID] = paywallInfo{
+		address:     user.NewUserPaywallAddress,
+		amount:      user.NewUserPaywallAmount,
+		txNotBefore: user.NewUserPaywallTxNotBefore,
+	}
+}
+
+func (b *backend) derivePaywallInfo(user *database.User) (string, uint64, int64, error) {
+	address, err := util.DerivePaywallAddress(b.params,
+		b.cfg.PaywallXpub, uint32(user.ID))
+	if err != nil {
+		err = fmt.Errorf("Unable to derive paywall address #%v "+
+			"for %v: %v", uint32(user.ID), user.Email, err)
+	}
+
+	return address, b.cfg.PaywallAmount, time.Now().Unix(), err
+}
+
+func (b *backend) checkForPayments() {
+	minConfirmations := b.cfg.MinConfirmationsRequired
+
+	mutex.Lock()
```

Comment:
That is a lot of locking gymnastics. If these are relatively fast operations just take the lock and defer the unlock. If they are slow then we need to do some gymnastics. Like break the function apart and do the bits that need locking in a single function; reason for that is that lcoking doesn't become intertwined like it is now. If someone adds a return we'll have a deadlock.

Can we do the init call in a more generic init area? pretty sure we shouldn't be calling the init all the time.